torchdynamo/convert_frame.py

```diff
@@ -316,6 +319,9 @@ def format_guard_failures(code):
         if not has_tensor_in_frame(frame):
             return None
 
+        global initial_grad_state
+        initial_grad_state = torch.is_grad_enabled()
```

Comment:
It's odd to me that we special case this, and to @jansel's point, not great that it's global - what we should have instead is a state snapshot that is passed around. Grad is just one such state.

Reply:
I'll address the global variable issue in a follow up PR. We are planning to publish the E2E training performance result, which is blocked by this PR.